### PR TITLE
Update performance of global styles code

### PR DESCRIPTION
**Pull request by oandregal**
This PR improves the performance of global styles by not looking up for resources we don't need.

What this means in practice:

- Styles: we no longer enqueue the global styles stylesheet if a theme doesn't have theme.json. We used to do that to make the presets available (both CSS Custom Properties & classes) for all themes automatically. People have reported this is not ideal and may interfere with their themes' CSS.

- Settings: whether or not the theme has a theme.json, we need to enqueue the settings to the editors to control design tools visibility, etc. This is now bounded to not require user data lookup if it has no support.

### Test

- Themes that support FSE & Global Styles work as expected ― theme example: TT1-blocks
  - site-editor: changes in the user styles are reflected in the front-end
  - post-editor: still works as expected (presets are present, etc)

- Themes that don't support FSE and/or Global Styles work as expected ― theme example: TwentyTwentyOne, TwentyTwenty.
  - site-editor: is not loaded for them
  - post-editor: still works as expected (presets are present, etc)

### CODE REVIEW

File: lib/class-wp-theme-json-resolver.php [hunk 1]
```diff
@@ -270,6 +274,10 @@ private static function get_user_origin() {
 	 * @return WP_Theme_JSON
 	 */
 	public function get_origin( $theme_support_data = array(), $origin = 'user', $merged = true ) {
+		if ( ! gutenberg_experimental_global_styles_has_theme_json_support() ) {
```
Comment (jorgefilipecosta) [inline lib/class-wp-theme-json-resolver.php:277]: I'm not sure if should be this function verifying if theme.json is supported or not. If this function is called with an origin user  I guess it should return the origin user. The caller of this function should probably be the one making the checks?

File: lib/class-wp-theme-json-resolver.php [hunk 2]
```diff
@@ -180,6 +180,10 @@ private function get_theme_origin( $theme_support_data = array() ) {
 	 * @return array Custom Post Type for the user's origin config.
 	 */
 	private static function get_user_data_from_custom_post_type( $should_create_cpt = false, $post_status_filter = array( 'publish' ) ) {
+		if ( ! gutenberg_experimental_global_styles_has_theme_json_support() ) {
```
Comment (jorgefilipecosta) [inline lib/class-wp-theme-json-resolver.php:183]: The same comment as in get_origin, it seems this function should always return the user custom_post_type and the checks should not be here.

I guess the checks should be in gutenberg_experimental_global_styles_enqueue_assets and gutenberg_experimental_global_styles_settings?

File: lib/global-styles.php [hunk 3]
```diff
@@ -235,6 +243,9 @@ function_exists( 'gutenberg_is_edit_site_page' ) &&
 	return $settings;
 }
 
-add_action( 'init', array( 'WP_Theme_JSON_Resolver', 'register_user_custom_post_type' ) );
+if ( gutenberg_experimental_global_styles_has_theme_json_support() ) {
```
Comment (jorgefilipecosta) [inline lib/global-styles.php:246]: I think the function gutenberg_experimental_global_styles_has_theme_json_support  must be called inside actions. Right now with this change, I'm getting some notices:
```

  | <br />
-- | --
  | <b>Notice</b>:  Use of undefined constant STYLESHEETPATH - assumed 'STYLESHEETPATH' in <b>/Users/pc/dev/core/wordpress-develop/build/wp-includes/template.php</b> on line <b>663</b><br />
  | <br />
  | <b>Notice</b>:  Use of undefined constant TEMPLATEPATH - assumed 'TEMPLATEPATH' in <b>/Users/pc/dev/core/wordpress-develop/build/wp-includes/template.php</b> on line <b>666</b><br />
  | <br />
  | <b>Warning</b>:  Cannot modify header information - headers already sent by (output started at /Users/pc/dev/core/wordpress-develop/build/wp-includes/template.php:663) in <b>/Users/pc/dev/core/wordpress-develop/build/wp-includes/functions.php</b> on line <b>6360</b><br />
  | <br />
  | <b>Warning</b>:  Cannot modify header information - headers already sent by (output started at /Users/pc/dev/core/wordpress-develop/build/wp-includes/template.php:663) in <b>/Users/pc/dev/core/wordpress-develop/build/wp-admin/includes/misc.php</b> on line <b>1310</b><br />
  | <br />
  | <b>Warning</b>:  Cannot modify header information - headers already sent by (output started at /Users/pc/dev/core/wordpress-develop/build/wp-includes/template.php:663) in <b>/Users/pc/dev/core/wordpress-develop/build/wp-admin/admin-header.php</b> on line <b>9</b><br />

```

I think it happens because some constants required are not initialized during this code execution.

**Comment by jorgefilipecosta** [comment]
I think we are causing a regression, in themes with a theme.json file e.g: Twenty Twenty-One Blocks we are outputting the expected styles.

**Comment by nosolosw** [comment]
@jorgefilipecosta your feedback was addressed.

File: lib/global-styles.php [hunk 4]
```diff
@@ -162,6 +162,10 @@ function gutenberg_experimental_global_styles_get_stylesheet( $tree, $type = 'al
  * and enqueues the resulting stylesheet.
  */
 function gutenberg_experimental_global_styles_enqueue_assets() {
+	if ( ! gutenberg_experimental_global_styles_has_theme_json_support() ) {
```
Comment (jorgefilipecosta) [inline lib/global-styles.php:165]: One of the advantages global styles brings is that themes don't need to create preset classes anymore. With this change I guess we remove this advantage right? I guess if theme.json is not present we don't need to make database queries and retrieve user settings but I guess we can still output preset classes?

**Review by jorgefilipecosta — COMMENTED** [review]
It seems in this branch the color palettes from the themes are not being applied and the default ones are always being used. I tested with the 2020 theme.

File: lib/global-styles.php (hunk 4 above)
Comment (nosolosw) [reply lib/global-styles.php:165]: Yes, one of the goals of this PR is that we will no longer enqueue any CSS **if the theme doesn't have theme.json support**. This is a mechanism to make sure that 1) we don't interfere with the theme's CSS and 2) we don't load unnecessary styles. The current behavior has been reported as not ideal in core-editor by people. I still think it's valuable if the theme supports theme.json, though.

**Comment by nosolosw** [comment]
> It seems in this branch the color palettes from the themes are not being applied and the default ones are always being used. I tested with the 2020 theme.

oh! good catch, thank you.

File: lib/class-wp-theme-json.php [hunk 5]
```diff
@@ -337,6 +339,7 @@ public function __construct( $contexts = array() ) {
 			// Process settings subtree.
 			$this->process_key( 'settings', $context, self::SCHEMA );
 			if ( isset( $context['settings'] ) ) {
+				$this->process_key( 'border', $context['settings'], self::SCHEMA['settings'] );
```
Comment (nosolosw) [inline lib/class-wp-theme-json.php:342]: This is unrelated and should have been added in the PR that added the border as settings, but I'm adding here.

File: lib/class-wp-theme-json.php [hunk 6]
```diff
@@ -324,8 +325,9 @@ public function __construct( $contexts = array() ) {
 			// Process styles subtree.
 			$this->process_key( 'styles', $context, self::SCHEMA );
 			if ( isset( $context['styles'] ) ) {
-				$this->process_key( 'color', $context['styles'], self::SCHEMA['styles'] );
-				$this->process_key( 'typography', $context['styles'], self::SCHEMA['styles'] );
+				$this->process_key( 'color', $context['styles'], self::SCHEMA['styles'], $should_escape_styles );
+				$this->process_key( 'spacing', $context['styles'], self::SCHEMA['styles'], $should_escape_styles );
```
Comment (nosolosw) [inline lib/class-wp-theme-json.php:329]: This is unrelated but I've noticed that was missing, so I'm adding it here.

File: lib/global-styles.php [hunk 7]
```diff
@@ -196,45 +200,74 @@ function gutenberg_experimental_global_styles_settings( $settings ) {
...
 
 	// STEP 1: ADD FEATURES
-	// These need to be added to settings always.
-	$settings['__experimentalFeatures'] = $all->get_settings();
+	//
+	// These need to be always added to the editor settings,
+	// even for themes that don't support theme.json.
+	// An example of this is that the presets are configured
+	// from the theme support data.
+	$settings['__experimentalFeatures'] = $tree->get_settings();
```
Comment (nosolosw) [inline lib/global-styles.php:219]: This will still look up for the theme.json, although the theme doesn't have one. I'd like to squeeze this performance bit as well, but it was getting convoluted so I'd like to do it in a follow-up (it requires improving the `WP_Theme_JSON_Resolver` API for consumers). So, this PR improves a bit the situation but doesn't fix all the things.

File: lib/global-styles.php [hunk 8]
```diff
@@ -196,45 +200,74 @@ function gutenberg_experimental_global_styles_settings( $settings ) {
...
+	// needs the ability to create them. Hence, we pass it some data
+	// for this: base styles (core+theme) and the ID of the user CPT.
 	$screen = get_current_screen();
 	if (
 		! empty( $screen ) &&
 		function_exists( 'gutenberg_is_edit_site_page' ) &&
 		gutenberg_is_edit_site_page( $screen->id ) &&
-		gutenberg_experimental_global_styles_has_theme_json_support()
+		gutenberg_experimental_global_styles_has_theme_json_support() &&
+		gutenberg_is_fse_theme()
```
Comment (nosolosw) [inline lib/global-styles.php:232]: We have a couple of checks that are related but independent:

- whether a theme is FSE-enabled (has a `block-templates/index.html`) => this is the check we use to decide whether to load the site editor, hence we also check for this when deciding whether to lookup user styles for the site editor.
- whether a theme is global-styles enabled (has an `experimental-theme.json`) => if a theme doesn't have theme.json it doesn't make sense to offer user styles, as 1) this are merged with the precedent origins (core, theme) and 2) we don't know if the user styles will play well with the theme's CSS.

**Review by jorgefilipecosta — APPROVED** [review]
LGTM 👍

**Comment by hoangitt** [comment]
Hello team, 
I'm using wordpress 5.6. I created file theme.json but not worked in my site
Structure JSON FILE:

`{
    "global": {
        "settings": {
            "custom": {
                "typo": {
                    "rootSize": 16,
                    "scale": "var(--viweb-paragraph-font-size-scale)"
                },
                "line-height": {
                    "small": 1.2,
                    "medium": 1.4,
                    "large": 1.8
                }
            },
            "spacing": {
                "customPadding": true,
                "units": ["px", "em", "rem", "vh", "vw"]
            },
            "color": {
                "palette": [
                    {
                        "slug": "primary",
                        "color": "var(--viweb-palette-primary-color)"
                    },
                    {
                        "slug": "second",
                        "color": "var(--viweb-palette-second-color)"
                    },
                    {
                        "slug": "orange",
                        "color": "var(--viweb-palette-orange-color)"
                    },
                    {
                        "slug": "blue",
                        "color": "var(--viweb-palette-blue-color)"
                    }
                ]
            },
            "typography": {
                "fontSizes": [
                    {
                        "slug": "small",
                        "size": "calc(var(--viweb-paragraph-font-size-small) / var(--viweb-paragraph-font-size-scale))"
                    },
                    {
                        "slug": "medium",
                        "size": "calc(var(--viweb-paragraph-font-size-medium) * 1px)"
                    },
                    {
                        "slug": "large",
                        "size": "calc(var(--viweb-paragraph-font-size-large) * var(--viweb-paragraph-font-size-scale))"
                    },
                    {
                        "slug": "huge",
                        "size": "calc(var(--viweb-paragraph-font-size-huge) * var(--viweb-paragraph-font-size-scale))"
                    }
                ],
                "customFontSize": true,
                "customLineHeight": true
            }
        }
    },
    "core/heading/h1": {
        "styles": {
            "typography": {
                "fontSize": "var(--viweb-heading-h1-font-size)"
            }
        }
    },
    "core/heading/h2": {
        "styles": {
            "typography": {
                "fontSize": "var(--viweb-heading-h2-font-size)"
            }
        }
    },
    "core/heading/h3": {
        "styles": {
            "typography": {
                "fontSize": "var(--viweb-heading-h3-font-size)"
            }
        }
    },
    "core/heading/h4": {
        "styles": {
            "typography": {
                "fontSize": "var(--viweb-heading-h4-font-size)"
            }
        }
    },
    "core/heading/h5": {
        "styles": {
            "typography": {
                "fontSize": "var(--viweb-heading-h5-font-size)"
            }
        }
    },
    "core/heading/h6": {
        "styles": {
            "typography": {
                "fontSize": "var(--viweb-heading-h6-font-size)"
            }
        }
    }
}`

**Comment by nosolosw** [comment]
@hoangitt to use the `experimental-theme.json` you need to use the Gutenberg plugin, it wasn't released as part of WordPress 5.6.

**Comment by hoangitt** [comment]
@nosolosw yep, thanks for reply
